### PR TITLE
add popup damage

### PR DESCRIPTION
**Pull request by shierote**
## Context

#116, #122 

## Summary

- Added xdg_shell's popup damage.
- This allows, for example, right-clicking on a wayland client to display/hide the popup. Details of what can now be done are added in the "How to check behavior" section.
- Added view_child struct to handle damage.

## How to check behavior
The following behaviors can be checked with the Wayland clients (e.g., `weston-terminal` `weston-flower`)
- showing popup
  - When you right click on the view, a popup appears around the cursor.
    - The requirement that if the cursor is around the edge of the screen, the popup appear inside the screen, is todo.
  - When you right click anywhere else, the previous popup disappears and a new popup appears around the cursor.
- hiding popup
  - (right/left) click elsewhere in the view
  - close the view
  - start `move`
  - focus/restack another view

- The behavior related to board is not defined.
  - When the view is focused during the board operation, the popup will remain.
  - When the view is unfocused during the board operation, the popup will disappear.

NOTE:
Below are the TODOs.
- popup's popup (popup tree)
  - handle the case that view_child's parent(view_child) exists.
- if the cursor is around the edge of the screen, the popup appear inside the screen.
- (subsurface)

### CODE REVIEW

File: zen/scene/view-child.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+{
+  struct wlr_surface *surface = self->impl->get_wlr_surface(self);
+  struct wlr_fbox damage_box;
+  pixman_region32_t damage;
+  pixman_box32_t *rects;
+  int rect_count;
+  int sx, sy;
+
+  if (!zn_view_child_is_mapped(self) || self->view == NULL ||
+      self->view->board == NULL || self->view->board->screen == NULL)
```
Comment (Aki-7) [inline zen/scene/view-child.c:26]: `self->view->board` is non null.

**Review by Aki-7 — CHANGES_REQUESTED** [review]

File: zen/xdg-toplevel-view.c [hunk 2]
```diff
@@ -174,11 +205,12 @@ zn_xdg_toplevel_view_create(
...
 zn_xdg_toplevel_view_destroy(struct zn_xdg_toplevel_view* self)
 {
-  wl_list_remove(&self->wlr_xdg_surface_destroy_listener.link);
   wl_list_remove(&self->wlr_surface_commit_listener.link);
+  wl_list_remove(&self->wlr_xdg_surface_destroy_listener.link);
   wl_list_remove(&self->move_listener.link);
-  wl_list_remove(&self->map_listener.link);
+  wl_list_remove(&self->new_popup_listener.link);
   wl_list_remove(&self->unmap_listener.link);
+  wl_list_remove(&self->map_listener.link);
```
[the remark this answers is not in the document]
Comment (shierote) [reply zen/xdg-toplevel-view.c:213]: This has been replaced to remove the members of `struct zn_xdg_toplevel_view` (`include/zen/xdg-toplevel-view.h`) in the reverse order of its definition. There is no technical reason.

File: include/zen/xdg-popup.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#include <wlr/types/wlr_xdg_shell.h>
+
+#include "zen/scene/view-child.h"
+
+struct zn_xdg_popup {
+  struct zn_view_child base;
+  struct wlr_xdg_popup *wlr_xdg_popup;  // nonnull
+
+  struct wl_listener map_listener;
+  struct wl_listener unmap_listener;  //
```
[the remark this answers is not in the document]
Comment (shierote) [reply include/zen/xdg-popup.h:12]: thx!

File: zen/scene/view-child.c [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  self->view_unmap_listener.notify = zn_view_child_view_unmap_handler;
+  wl_signal_add(&self->view->events.unmap, &self->view_unmap_listener);
+  self->view_destroy_listener.notify = zn_view_child_view_destroy_handler;
+  wl_signal_add(&self->view->events.destroy, &self->view_destroy_listener);
+}
+
+void
+zn_view_child_fini(struct zn_view_child *self)
+{
+  if (zn_view_child_is_mapped(self)) zn_view_child_damage(self);
```
[the remark this answers is not in the document]
Comment (shierote) [reply zen/scene/view-child.c:117]: That seems simple and nice to me!

File: zen/scene/view-child.c [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+static void
+zn_view_child_view_destroy_handler(struct wl_listener *listener, void *data)
+{
+  struct zn_view_child *self =
+      zn_container_of(listener, self, view_destroy_listener);
+  UNUSED(data);
+
+  zn_view_child_damage(self);
+  self->mapped = false;
```
[the remark this answers is not in the document]
Comment (shierote) [reply zen/scene/view-child.c:93]: Are you implying that I should delete the `self->mapped = false` line?

[the remark this answers is not in the document]
Comment (Aki-7) [reply zen/scene/view-child.c:93]: Sorry, I wanted to say `zn_view_chile_unmap`.
```suggestion
  zn_view_child_unmap(self);
```

**Review by watasuke102 — CHANGES_REQUESTED** [review]

File: zen/xdg-toplevel-view.c [hunk 6]
```diff
@@ -51,13 +54,26 @@ zn_xdg_toplevel_view_unmap(struct wl_listener* listener, void* data)
...
+
   wl_list_remove(&self->move_listener.link);
   wl_list_init(&self->move_listener.link);
 
   wl_list_remove(&self->wlr_surface_commit_listener.link);
   wl_list_init(&self->wlr_surface_commit_listener.link);
 }
 
+static void
+zn_xdg_toplevel_view_new_popup(struct wl_listener* listener, void* data)
```
[the remark this answers is not in the document]
Comment (shierote) [reply zen/xdg-toplevel-view.c:68]: This follows other code conventions.
However, I agree that it is difficult to recognize it as a handler, so I will rename the other parts of the code as well. Thanks.

File: zen/scene/view-child.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (shierote) [reply zen/scene/view-child.c:93]: I understand! Good idea, thanks!!

File: zen/xdg-toplevel-view.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (shierote) [reply zen/xdg-toplevel-view.c]: I unified all handler names with *_handler in https://github.com/zigen-project/zen/pull/157/commits/b549b1e28fb42e8face57132df69ac6397befe03. check it out.

**Comment by shierote** [comment]
I fixed all review points. thx!
Also, I took this opportunity to unify all handler names with *_handler in https://github.com/zigen-project/zen/commit/b549b1e28fb42e8face57132df69ac6397befe03. check it out.

**Comment by Aki-7** [comment]
It's me who started to use `*_handler`, but now I think `handle_*` is better, because in the object oriented programming, the method should be basically verbs.
Could you rewrite b549b1e with `handle_*`? It's very good timing to fix it.

**Comment by shierote** [comment]
@Aki-7 
> It's me who started to use *_handler, but now I think handle_* is better, because in the object oriented programming, the method should be basically verbs.

I fixed it.

**Review by watasuke102 — APPROVED** [review]
_LGTM_

**Review by Aki-7 — APPROVED** [review]
Looks good to me!

**Pull request merged**